File: problems/栈与队列/155.最小栈.py
#!/usr/bin/env python3
"""
155. 最小栈
算法: 双栈（主栈+辅助最小栈）
时间复杂度: O(1)（push/pop/top/getMin均为O(1)）
空间复杂度: O(n)
"""
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if self.stack.pop() == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


if __name__ == "__main__":
    min_stack = MinStack()
    min_stack.push(-2)
    min_stack.push(0)
    min_stack.push(-3)
    print(min_stack.getMin())  # 预期: -3
    min_stack.pop()
    print(min_stack.top())     # 预期: 0
    print(min_stack.getMin())  # 预期: -2

File: problems/栈与队列/155.最小栈.cpp
#include <iostream>
#include <stack>
using namespace std;

class MinStack {
public:
    stack<int> st, minSt;

    void push(int val) {
        st.push(val);
        if (minSt.empty() || val <= minSt.top()) minSt.push(val);
    }

    void pop() {
        if (st.top() == minSt.top()) minSt.pop();
        st.pop();
    }

    int top() { return st.top(); }

    int getMin() { return minSt.top(); }
};

int main() {
    // 测试用例
    MinStack ms;
    ms.push(-2);
    ms.push(0);
    ms.push(-3);
    cout << ms.getMin() << endl;  // 预期: -3
    ms.pop();
    cout << ms.top() << endl;     // 预期: 0
    cout << ms.getMin() << endl;  // 预期: -2

    return 0;
}
